[oracle] Enhancement to capture system statistics metrics, pga metrics, sga metrics

agithomas opened the pull request:
Enhancement to capture system statistics metrics, pga metrics, sga metrics. 

## Type of change

- Enhancement


## What does this PR do?

- Enhancement to capture system statistics metrics, pga metrics, sga metrics. 
- Added dashboard to capture performance metrics, system statistics metrics, pga and sga metrics.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## How to test this PR locally

- `docker run -d --name oracle -p 1521:1521 -p 5500:5500 docker.elastic.co/observability-ci/database-enterprise:12.2.0.1`
- Install oracle client in the host / vm
- elastic-package stack up --version 8.3.0
- Access kibana, login, configure oracle integration. Use the default values

## Related issues

- Relates #3911 

## Screenshots
<img width="1663" alt="Screenshot 2022-08-08 at 12 24 53 PM" src="https://user-images.githubusercontent.com/101976829/183357706-f8236b29-2cae-45f6-a2ac-bdd1bfed83c4.png">
<img width="1655" alt="Screenshot 2022-08-08 at 12 24 32 PM" src="https://user-images.githubusercontent.com/101976829/183357731-9fe74172-5117-40b5-900b-ed010d0801b0.png">
<img width="1655" alt="Screenshot 2022-08-08 at 12 24 14 PM" src="https://user-images.githubusercontent.com/101976829/183357734-31626373-f6bf-4241-b619-c70d2febc337.png">

elasticmachine commented:
## :broken_heart: Tests Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3967/detail/PR-3967/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3967/detail/PR-3967/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3967/detail/PR-3967/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3967/detail/PR-3967/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://integrations_3967.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-08-08T06:45:59.255Z&rangeTo=2022-08-08T07:05:59.255Z&transactionName=Ingest-manager/integrations/PR-{number}&transactionType=job&latencyAggregationType=avg&traceId=ec3bc29ecbf56aa88aa1382cd1a26ac3&transactionId=c47636d7f0c65e68)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-08-08T06:55:59.255+0000


* Duration: 15 min 53 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 1  |
| Passed       | 20  |
| Skipped      | 0 |
| Total        | 21   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->

  ### Test errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3967/detail/PR-3967/1//tests)
<details><summary>Expand to view the tests failures</summary><p>
    
    
##### `static test: Verify sample_event.json – oracle.pga_sga`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 null 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 one or more errors found in document: [0] field "oracle.pga_sga.numeric.total pga used for auto workareas" is undefined
[1] field "oracle.pga_sga.numeric.cache hit percentage" is undefined
[2] field "oracle.pga_sga.numeric.total freeable pga memory" is undefined
[3] field "oracle.pga_sga.numeric.total pga inuse" is undefined
[4] field "oracle.pga_sga.numeric.global memory bound" is undefined
[5] field "oracle.pga_sga.numeric.aggregate pga auto target" is undefined
[6] field "oracle.pga_sga.numeric.aggregate pga target parameter" is undefined
[7] field "oracle.pga_sga.numeric.maximum pga allocated" is undefined
[8] field "oracle.pga_sga.numeric.process count" is undefined
[9] field "oracle.pga_sga.numeric.total pga allocated" is undefined
[10] field "oracle.pga_sga.numeric.max processes count" is undefined 
 ```
</p></details>
</ul>
    
</p></details>


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3967/detail/PR-3967/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Test integration: oracle`
<ul>
<li>Took 1 min 55 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-3967/runs/1/steps/921/log/?start=0">here</a></li>
<li>Description: <code>                                eval "$(../../build/elastic-package stack shellinit)"                                ../../build/elastic-package test -v --report-format xUnit --report-output file --test-coverage                                </code></l1>
</ul>
  
  </p>
  </details>



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 100.0% (`1/1`) | :green_heart: 
Files | 100.0% (`1/1`) | :green_heart: 2.918
Classes | 100.0% (`1/1`) | :green_heart: 2.918
Methods | 82.759% (`24/29`) | :-1: -6.798
Lines | 89.13% (`164/184`) | :-1: -2.448
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

efd6 commented on packages/oracle/data_stream/performance/fields/fields.yml:
```diff
@@ -117,3 +117,48 @@
+      type: double
+      metric_type: gauge
+      description: This metric checks for failed DBMS jobs.
+    - name: sessioncount
```
Maybe `s/sessioncount/session_count/`? (with associated changes throughout)

ishleenk17 commented on packages/oracle/data_stream/system_statistics/fields/fields.yml:
```diff
@@ -0,0 +1,333 @@
+  fields:
+    - name: parallel_operations_not_downgraded
+      type: double
+      metric_type: counter
```
There are approx ~80 metrics for this datastream. 
Maybe we can reduce it to top 50. 
Thinking from perspective if customer would require all of these ?

agithomas replied:
dynamic_metric_name_filter is  available to limit the data. You may refer the manifest.yml

lalit-satapathy replied:
@agithomas where is the dynamic filtering exactly, Can you point to it.

agithomas replied:
For the query, refer packages/oracle/data_stream/system_statistics/agent/stream/stream.yml.hbs

ishleenk17 reviewed (commented):
ECS and base fields not present in performance data stream

ishleenk17 reviewed (commented):
IIRC, SQL has flexibility of letting the user choose which field they want to retrive from a table in DB. 
Would that be inherited by oracle as well since its using sql metrics ?

agithomas commented:
> IIRC, SQL has flexibility of letting the user choose which field they want to retrive from a table in DB. Would that be inherited by oracle as well since its using sql metrics ?

dynamic_metric_name_filter feature is present to permit the user to configure based on his / her requirements

agithomas commented:
> ECS and base fields not present in performance data stream

It is present. But, you won't find in the Files Changed. We already have Performance data stream and this time we are extending it. base-fields.yml and ecs.yml files were added earlier.

agithomas commented:
@efd6 , all the issues highlighted are fixed. Kindly review

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > Either there was a build timeout or someone aborted the build.
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3967/detail/PR-3967/2//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3967/detail/PR-3967/2//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3967/detail/PR-3967/2//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3967/detail/PR-3967/2//artifacts)   [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2022-09-12T08:53:06.501Z&rangeTo=2022-09-12T09:13:06.501Z&transactionName=BUILD+Ingest-manager%2Fintegrations%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=43019eaa9263f399c85657c621a36bd1&transactionId=b26e70a83baea3c3)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-09-12T09:03:06.501+0000


* Duration: 15 min 8 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->
















## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

efd6 commented on packages/oracle/changelog.yml:
```diff
@@ -1,4 +1,9 @@
+  changes:
+    - description: Enhancement to capture system statistics metrics, pga metrics, sga metrics.
+      type: enhancement
+      link: https://github.com/elastic/integrations/pull/3911
```
```suggestion
      link: https://github.com/elastic/integrations/pull/3967
```

efd6 reviewed (commented):
After nit this LGTM. The oracle SQL should probably also be checked by another; it looks OK to this non-expert.

muthu-mps commented on packages/oracle/data_stream/performance/fields/fields.yml:
```diff
@@ -117,3 +117,48 @@
+          type: double
+          unit: percent
+          metric_type: gauge
+          description: Percentage of number of time waits of waits that are not Idle wait class.
```
Can we rephrase the more complex sentence into a simple readable one ?

muthu-mps commented on packages/oracle/data_stream/performance/fields/fields.yml:
```diff
@@ -117,3 +117,48 @@
+          description: Amount of time spent in the wait class by the session.
+        - name: total_waits
+          type: double
+          metric_type: gauge
```
If this is an incremental value then,Can we need to have the metric_type as counter ?
Does the `unit type: s` appropriate for this field ? As per the field name and description it seems to be a total count value.

muthu-mps commented on packages/oracle/data_stream/performance/fields/fields.yml:
```diff
@@ -117,3 +117,48 @@
           metric_type: gauge
           description: |
             Ratio of session cursor cache hits from total number of cursors.
+    - name: failed_db_jobs
```
The new changes are not reflected in the sample-even.json. Can we regenerate the events for performance ?

agithomas replied:
Performance DS generate 7 different events run. 
The current sample is from the 3rd query configured in the stream.yml.

muthu-mps commented on packages/oracle/data_stream/system_statistics/fields/fields.yml:
```diff
@@ -0,0 +1,333 @@
+      type: double
+      metric_type: counter
+      description: Total number of waits that occurred during an enqueue convert or get because the enqueue get was deferred.
+    - name: redo_size
```
unit:byte is not added for this field. This applies for below fields which has byte representation.

muthu-mps commented:
Can we realign the dashboard [packages/oracle/img/Oracle-system_statistics-dashboard.png](https://github.com/elastic/integrations/pull/3967/files#diff-ce2b8f9d39b2eed7f7348044bc8705a3054529a3bcd3e4dcf3b6b77833942fbd) and remove the extra gap after Inactive session count > 1hr.

muthu-mps commented on packages/oracle/data_stream/system_statistics/fields/fields.yml:
```diff
@@ -0,0 +1,333 @@
+    - name: transaction_lock_background_get_time
+      type: double
+      metric_type: counter
+      description: Useful only for internal debugging purposes.
```
It's not clear when we say it as only for internal debugging purpose. Can we remove this field if it's not useful ?

muthu-mps commented on packages/oracle/data_stream/pga_sga/fields/fields.yml:
```diff
@@ -0,0 +1,67 @@
+      description: Number of bytes of PGA memory in all processes that could be freed back to the operating system.
+      unit: byte
+      metric_type: gauge
+    - name: cache_hit_percentage
```
Should it be `_pct` as per the standard naming conventions ?

muthu-mps commented on packages/oracle/data_stream/pga_sga/fields/fields.yml:
```diff
@@ -0,0 +1,67 @@
+      unit: byte
+      description: Current amount of PGA memory allocated by the instance.
+      metric_type: gauge
+    - name: total_pga_used_for_auto_workareas
```
Can we make this field name with appropriate naming conventions ?

agithomas replied:
@muthu-mps , unable to figure out where it is inappropriate. Can you guide?

muthu-mps replied:
IMO - The field `total_pga_used_for_auto_workareas` `used_for` can be removed and the field name could be renamed. I am okay if the field follows the naming standards.

agithomas replied:
These are key names / keywords of name:value pair in oracle database views. So, keeping them without much change, after internal discussion.

lalit-satapathy commented on packages/oracle/data_stream/performance/agent/stream/stream.yml.hbs:
```diff
@@ -14,4 +14,12 @@ sql_queries:
+  - query: select WAIT_CLASS, TOTAL_WAITS, round(100 * (TOTAL_WAITS / SUM_WAITS),2) PCT_WAITS, ROUND((TIME_WAITED / 100),2) TIME_WAITED_SECS, round(100 * (TIME_WAITED / SUM_TIME),2) PCT_TIME from (select WAIT_CLASS, TOTAL_WAITS, TIME_WAITED from V$SYSTEM_WAIT_CLASS where WAIT_CLASS != 'Idle'), (select sum(TOTAL_WAITS) SUM_WAITS, sum(TIME_WAITED) SUM_TIME from V$SYSTEM_WAIT_CLASS where WAIT_CLASS  ...
+    response_format: table
+
+  
```
How many documents are created by these queries? Should we add  [request_merge](request_merge) flag? This question applies to all the oracle data streams.

agithomas replied:
pga_sga : 2
sysmetric : 1
system_statistics : 1

There is a scope of improvement in performance datastream (part of 8.4) which has 7 queries.

agithomas replied:
As per the discussion, a separate PR will be created for this as 8.4.0 features are needed, changing min kibana version to 8.4.0 for this integration

lalit-satapathy commented on packages/oracle/data_stream/performance/fields/fields.yml:
```diff
@@ -117,3 +117,47 @@
+          unit: percent
+          metric_type: gauge
+          description: Percentage of time waits that are not Idle wait class.
+        - name: pct_waits
```
should we remove "wait" for last field names. Need to think if having these in the names better since the group name is "wait"?

agithomas replied:
```
- name: pct_time
          type: double
          unit: percent
          metric_type: gauge
          description: Percentage of time waits that are not Idle wait class.
        - name: pct_waits
          type: double
          unit: percent
          metric_type: gauge
          description: Percentage of number of pct time waits that are not of Idle wait class.
```

Cannot remove `wait` because, it will lead to conflicts / confusions

agithomas commented:
> Can we realign the dashboard [packages/oracle/img/Oracle-system_statistics-dashboard.png](https://github.com/elastic/integrations/pull/3967/files#diff-ce2b8f9d39b2eed7f7348044bc8705a3054529a3bcd3e4dcf3b6b77833942fbd) and remove the extra gap after Inactive session count > 1hr.

Corrected.

lalit-satapathy commented on packages/oracle/data_stream/performance/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -83,6 +83,83 @@ processors:
+        ignore_failure: true
+    - rename:
+        field: oracle.performance.pct_time
+        target_field: oracle.performance.wait.pct_time
```
Can we make sure these oracle.performance.wait.* fields are there in the sample events? also I am not able to locate them in the README file also.

muthu-mps commented on packages/oracle/data_stream/performance/fields/fields.yml:
```diff
@@ -117,3 +117,47 @@
+    - name: session_count
+      type: group
+      field:
+        - name: active_session_count
```
Group name has session_count. Consider removing session_count suffix from active and inactive fields.

muthu-mps commented on packages/oracle/data_stream/performance/fields/fields.yml:
```diff
@@ -117,3 +117,47 @@
+        - name: inactive_session_count
+          type: double
+          metric_type: gauge
+          description: Total inactive sessions more than one hour.
```
Do you think that the description here should be exchanged with the above field ?

muthu-mps commented on packages/oracle/data_stream/system_statistics/fields/fields.yml:
```diff
@@ -0,0 +1,335 @@
+      type: double
+      metric_type: counter
+      description: Total number of calls (user and recursive) that executed SQL statements.
+    - name: process_last_non-idle_time
```
Consider using only underscores and not hyphen in field names.

muthu-mps commented on packages/oracle/data_stream/pga_sga/fields/fields.yml:
```diff
@@ -0,0 +1,67 @@
+- name: oracle.pga_sga
```
Is it okay to rename this datastream to `memory`  ? As we have the metrics related to memory and also the two data points pga and sga can be grouped separately.
@lalit-satapathy - Please share your inputs on the same.

agithomas replied:
PGA and SGA are specifically called out memory areas. Hence, not generalised as `memory`

A Program Global Area (PGA) is a memory region that contains data and control information for a server process. It is nonshared memory created by Oracle Database when a server process is started. Access to the PGA is exclusive to the server process.

The System Global Area (SGA) is a group of shared memory structures that contain data and control information for one Oracle Database instance.

It is different from DB instance memory.

muthu-mps replied:
I am just considering, Is it possible to group pga and sga metrics separately ? That would give a clear view on `pga` and `sga` metrics.

agithomas replied:
Technically it is possible. But, i do not find it is needed. 

1. Each DS then has too few fields
2. Users who use PGA tend to have a look at SGA and vice versa.
3. Too few queries under sql_queries and limits possibility of combining events (planned for future).

agithomas replied:
Please refer the conversation : https://github.com/elastic/integrations/pull/3967#discussion_r971967269

lalit-satapathy replied:
> I am just considering, Is it possible to group pga and sga metrics separately ? That would give a clear view on `pga` and `sga` metrics.

The suggestion is to group the fields in the same data stream under "pga" and "sga" ? That will necessitate a rename of these fields?

agithomas replied:
@lalit-satapathy , yes rename processors are needed. Please confirm if to make grouping changes?

muthu-mps commented on packages/oracle/_dev/build/docs/README.md:
```diff
@@ -82,3 +82,15 @@ The `database_audit` dataset collects Oracle Audit logs.
 
 {{event "sysmetric"}}
+
+### PGA SGA Metrics 
```
Can we add few lines in the document for each datastream ?

elasticmachine commented:
### :rocket: Benchmarks report

#### Package `oracle` :+1:(0) :green_heart:(0) :broken_heart:(1)

<details>
<summary>Expand to view</summary>

Data stream | Previous EPS | New EPS | Diff (%) | Result
----------- | ------------ | ------- | -------- | ------
`database_audit` | 33333.33 | 25641.03 | -7692.3 (-30%) | :broken_heart:
</details>


To see the full report comment with `/test benchmark fullreport`

<!--COMMENT_GENERATED_WITH_ID_benchmark-report.md-->

muthu-mps commented on packages/oracle/_dev/build/docs/README.md:
```diff
@@ -82,3 +76,25 @@ The `database_audit` dataset collects Oracle Audit logs.
+
+### System Statistics Metrics 
+
+The System Global Area (SGA) is a group of shared memory structures that contain data and control information for one Oracle Database instance. Metrics concerning System Global Area (SGA) memory. Metrics concerning System Global Area (SGA) memory are mentioned below.
```
Please remove the repeated sentence - `Metrics concerning System Global Area (SGA) memory.`

muthu-mps commented on packages/oracle/data_stream/performance/agent/stream/stream.yml.hbs:
```diff
@@ -14,4 +14,12 @@ sql_queries:
+  - query: select WAIT_CLASS, TOTAL_WAITS, round(100 * (TOTAL_WAITS / SUM_WAITS),2) PCT_WAITS, ROUND((TIME_WAITED / 100),2) TIME_WAITED_SECS, round(100 * (TIME_WAITED / SUM_TIME),2) PCT_TIME from (select WAIT_CLASS, TOTAL_WAITS, TIME_WAITED from V$SYSTEM_WAIT_CLASS where WAIT_CLASS != 'Idle'), (select sum(TOTAL_WAITS) SUM_WAITS, sum(TIME_WAITED) SUM_TIME from V$SYSTEM_WAIT_CLASS where WAIT_CLASS  ...
+    response_format: table
+
+  
```
nit: Remove extra unused lines.

agithomas commented:
@lalit-satapathy , @muthu-mps , @efd6 ,

Can you please help reviewing the recent change and approve the PR?

From your last review, below changes are made

1. Modified datastream name from `pga_sga` to `memory`
2. Field mapping in datastream `memory` modified
3. Modified field name having `percentage` to `pct`
4. Added screenshots of memory metrics dashboard
5. README file wasn't displaying deep nested fields. This bug is fixed.
6. Typos fixed.

muthu-mps commented on packages/oracle/_dev/build/docs/README.md:
```diff
@@ -82,3 +76,25 @@ The `database_audit` dataset collects Oracle Audit logs.
+
+{{event "system_statistics"}}
+
+### Performance Metrics
```
While looking at the documentation in the preview page the Sysmetrics, Tablespace and Performance has no description for it. Do you think that we could add a one liner description for those ?

muthu-mps commented on packages/oracle/manifest.yml:
```diff
@@ -19,7 +48,7 @@ icons:
   - name: oracle
     title: Oracle Audit Logs
-    description: Collect Oracle Audit logs, Performance metrics, Tablespace metrics, Sysmetrics integration
+    description: Collect Oracle Audit logs, Performance metrics, Tablespace metrics, Sysmetrics integration, System statistics, memory metrics.
```
`memory metric M`. Can we make this to camel case ?

muthu-mps commented on packages/oracle/data_stream/memory/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,77 @@
+    - rename:
+        field: oracle.metrics
+        target_field: oracle.memory
+        ignore_missing: true
```
Do you consider adding ignore_failure for rest of the fields ? Because the other datastreams has ignore_failure for all the fields.

muthu-mps reviewed (approved):
LGTM!
